website/docs/user-defined-generics.md

```diff
@@ -0,0 +1,163 @@
+  BoxedType = type_member
+
+  sig { params(value: BoxedType).void }
+  def initialize(value)
+    @value = value
+  end
+
+  sig { returns(BoxedType) }
+  attr_accessor :value
+end
+
+# Omitted type members are set to T.untyped
+T.reveal_type(Box.new(3)) # Box[T.untyped]
+
+T.reveal_type(Box[Integer].new(3)) # Box[Integer]
+T.reveal_type(Box[Integer].new(3).value) # Integer
+
+T.reveal_type(Box[Integer].new("hello")) # error: Expected Integer but found String
+```
+
+You can define classes which accept multiple type parameters by using
+`type_member` several times, in order:
+
+```ruby
+class Pair
+  extend T::Sig
+  extend T::Generic
+
+  K = type_member
+  V = type_member
+
+  sig { params(key: K, value: V).void }
+  def initialize(key, value)
+    @key = key
+    @value = value
+  end
+
+  sig { returns(K) }
+  attr_accessor :key
+
+  sig { returns(V) }
+  attr_accessor :value
+end
+
+pair = Pair[String, Integer].new("foo", 2)
+T.reveal_type(pair) # Pair[String, Integer]
+T.reveal_type(pair.key) # String
+T.reveal_type(pair.value) # Integer
+```
+
+Using `[]` on a class would normally raise an exception, so extending
+`T::Generic` defines this as a method which simply returns your class.
+
+## Re-declaring Type Members
+
+If you inherit from a class, include a module, or extend a module which defines
+type members, then the type members must be re-declared in the child type.
+
+There are two ways to re-declare a type member:
+
+- **As a variant type member:** it must still be specified as a type parameter
+  when using the type (or left unspecified to use `T.untyped`), and will need to
+  be re-declared again in any child types.
+
+- **As a fixed type member:** it will take on a fixed value in this type, and
+  cannot be specified with a type parameter. Child types will need to re-declare
+  the type member, and this re-declaration must be fixed.
+
+```ruby
+class Base
+  extend T::Generic
+  A = type_member
+  B = type_member
+end
+base = Base[Integer, String].new
+
+class Subtype < Base
+  A = type_member
+  B = type_member
+end
+subtype = Subtype[Integer, String].new
+
+# B has been fixed, so now only A can be specified
+# Any signatures which use B will treat B as String
+class PartiallyFixedSubtype < Subtype
+  A = type_member
+  B = type_member(fixed: String)
+end
+partially_fixed_subtype = PartiallyFixedSubtype[Integer].new
+
+# Both types are fixed, so FullyFixedSubtype takes no type parameters
+# A will be treated as Integer, and B as String
+# (Even though the superclass fixed B to String, it must be fixed again)
+class FullyFixedSubtype < PartiallyFixedSubtype
+  A = type_member(fixed: Integer)
+  B = type_member(fixed: String)
+end
+fully_fixed_subtype = FullyFixedSubtype.new
+```
+
+## Variance
+
+Type members on modules and interfaces may be invariant (default), covariant, or
+contravariant.
+
+| Variance      | Definition          | Allowed usage in `sig` | Allowed types                |
```

Comment:
i don't understand what "Allowed usage in `sig`" means in this context, and the examples below don't seem to explain it. are you able to add an example for this?

Reply:
Ah, thanks - I missed the explanation of this. I've added another example at the bottom of the page :)

Reply:
Thanks so much for this. These docs are fantastic. Even if they don't get merged I'll be sharing this PR around with the team.

Reply:
Thank you very much! I hope they find these useful too :+1: